pkg/activator/net/revision_backends.go

```diff
@@ -215,14 +215,29 @@ func (rw *revisionWatcher) probeClusterIP(dest string) (bool, error) {
 // the ones that are successfully probed, whether the update was a no-op, or an error.
 // If probing fails but not all errors were compatible with being caused by
 // mesh being enabled, being enabled, notMesh will be true.
-func (rw *revisionWatcher) probePodIPs(dests sets.String) (succeeded sets.String, noop bool, notMesh bool, err error) {
-	// Short circuit case where dests == healthyPods
+func (rw *revisionWatcher) probePodIPs(ready, notReady sets.String) (succeeded sets.String, noop bool, notMesh bool, err error) {
+	dests := ready.Union(notReady)
+
+	// Short circuit case where all the current pods are already known to be healthy.
 	if rw.healthyPods.Equal(dests) {
 		return rw.healthyPods, true /*no-op*/, false /* notMesh */, nil
 	}
 
 	toProbe := sets.NewString()
 	healthy := sets.NewString()
+
+	if rw.meshMode != network.MeshCompatibilityModeAuto {
+		// If k8s marked the pod ready before we managed to probe it, and we're
+		// not also using the probe to sniff whether mesh is enabled, we can just
+		// trust k8s and mark it healthy without probing.
+		// TODO: replace with ready.Clone() once we have a recent-enough apimachinery (1.23+).
+		healthy = ready.Union(nil)
```

Comment:
Is the copying necessary? If so, why?

Reply:
I think so, because we muck with `healthy` a few lines down, and it's probably pretty unexpected that someone passing a set in to probePodIPs would find that set modified as a side effect (e.g. otherwise the "Failed probing pods" log message would end up logging a misleading set of ready pods).